src/Stratis.Bitcoin.Features.Consensus.Tests/TimeSyncBehaviorTest.cs

```diff
@@ -28,173 +28,283 @@ public class TimeSyncBehaviorTest
+            TimeSyncBehaviorState state = this.CreateTimeSyncBehaviorState(dateTimeProvider);
+
+            // add max inbound samples
+            for (int i = 0; i < TimeSyncBehaviorState.MaxInboundSamples; i++)
+            {
+                bool used = state.AddTimeData(new IPAddress(i), TimeSpan.FromSeconds(400), true);
                 Assert.True(used);
-                if (i >= outboundSamples - TimeSyncBehaviorState.MaxOutboundSamples) outSamples.Add(-i * 1000);
             }
 
-            var allSamples = new List<int>(inSamples);
-            for (int i = 0; i < TimeSyncBehaviorState.OutboundToInboundWeightRatio; i++)
-                allSamples.AddRange(outSamples);
+            // add enough outbound samples to get an adjustment - and across an even range
+            Assert.True(state.AddTimeData( new IPAddress(1), TimeSpan.FromSeconds(-500), false));
+            Assert.True(state.AddTimeData( new IPAddress(2), TimeSpan.FromSeconds(-250), false));
+            Assert.True(state.AddTimeData( new IPAddress(3), TimeSpan.FromSeconds(250), false));
+            Assert.True(state.AddTimeData( new IPAddress(4), TimeSpan.FromSeconds(500), false));
+
+            TimeSpan diff =  dateTimeProvider.GetAdjustedTime() -  dateTimeProvider.GetUtcNow();
+            Assert.True(Math.Abs(diff.TotalMilliseconds - 250*1000) < TimeEpsilonMs, $"should be 250ms because leaning towards the inbound value of 400ms, but still in outbound. Actual: {diff.TotalMilliseconds}ms");
+
+            // add another batch of inbound with a greater offset than highest outbound
+            for (int i = TimeSyncBehaviorState.MaxInboundSamples; i < TimeSyncBehaviorState.MaxInboundSamples*2; i++)
+            {
+                Assert.True(state.AddTimeData(new IPAddress(i*2), TimeSpan.FromSeconds(-1000), true), $"index: {i}");
+            }
+
+            diff =  dateTimeProvider.GetAdjustedTime() -  dateTimeProvider.GetUtcNow();
+            Assert.True(Math.Abs(diff.TotalMilliseconds - (-250*1000)) < TimeEpsilonMs, $"should be -250ms because the new inbound replace the old and swing the median closer to the new -1000ms values, but still in outbound. Actual: {diff.TotalMilliseconds}ms");
+        }
+         
+        /// <summary>
+        /// Sets up the system under test
+        /// </summary>
+        /// <param name="dateTimeProvider"></param>
+        /// <returns></returns>
+        private TimeSyncBehaviorState CreateTimeSyncBehaviorState(DateTimeProvider dateTimeProvider)
+        {
+            var lifetime = new NodeLifetime();
+            var loggerFactory = new LoggerFactory();
+            var asyncLoopFactory = new AsyncLoopFactory(loggerFactory);
+            var state = new TimeSyncBehaviorState(dateTimeProvider, lifetime, asyncLoopFactory, loggerFactory, Network.Main);
+            return state;
+        }
+    }
+
+    /// <summary>
+    /// Representation of a Sample for data driven tests.
+    /// </summary>
+    public class TestSample
```

Comment:
This class should be a subclass of TimeSyncBehaviorTest. Reason for that is that other tests would otherwise see it and they may have their own test sample class.

Reply:
This is not essential and could be done if it is ever needed. ie [Yagni](https://martinfowler.com/bliki/Yagni.html) - however since you have mentioned it, I'll do it.